### PR TITLE
Toffoli gates

### PR DESCRIPTION
**Pull request by Guillaumemartinfesta**
The implementation of Toffoli gates and the possibility to create multi controlled gates.

### CODE REVIEW

File: pyqtorch/modules/primitive.py [hunk 1]
```diff
@@ -427,3 +434,30 @@ def __init__(self, qubits: ArrayLike, n_qubits: int):
...
+
+        # Create a Toffoli gate with 3 control qubits.
+        toffoli_gate = pyq.Toffoli(qubits=[0, 1, 2, 3], n_qubits=4)
+
+        # Create a zero state
+        toffoli_state = pyq.zero_state(n_qubits=4)
+
+        # Apply the CSWAP gate to the zero state
+        result = toffoli_gate(toffoli_state)
+        print(result)
```
Comment (jpmoutinho) [inline pyqtorch/modules/primitive.py:461]: Typo on the CSWAP and I would change this example to initializing a one state just so the toffoli actually does something.

File: pyqtorch/modules/primitive.py [hunk 2]
```diff
@@ -427,3 +434,30 @@ def __init__(self, qubits: ArrayLike, n_qubits: int):
...
+            qubits (ArrayLike): The first n-1 qubits will be considered as the control
+            qubits and the last one will be the target qubit of the Toffoli gate.
+
+        Examples:
+        ```python exec="on" source="above" result="json"
+        import torch
+        import pyqtorch.modules as pyq
+
+        # Create a Toffoli gate with 3 control qubits.
+        toffoli_gate = pyq.Toffoli(qubits=[0, 1, 2, 3], n_qubits=4)
```
Comment (jpmoutinho) [inline pyqtorch/modules/primitive.py:454]: So is the target qubit always the final one on the qubit support list?

File: pyqtorch/core/operation.py [hunk 3]
```diff
@@ -40,20 +40,21 @@ def get_parametrized_matrix_for_operation(operation_type: str, theta: torch.Tens
...
 
     Returns:
 
         torch.Tensor: the resulting controlled gate populated by operation_matrix
     """
     mat_size = len(operation_matrix)
-    controlled_mat: torch.Tensor = torch.eye(2 * mat_size, dtype=torch.cdouble)
+    controlled_mat: torch.Tensor = torch.eye(2**n_control_qubits * mat_size, dtype=torch.cdouble)
     controlled_mat[-mat_size:, -mat_size:] = operation_matrix
     return controlled_mat
```
Comment (jpmoutinho) [inline pyqtorch/core/operation.py:59]: The docstring says it takes a `2x2`, but isn't it actually working for any size of the operation? That's what `mat_size` does right?

**Comment by nmheim** [comment]
Just took a super brief look but this looks alright to me

File: pyqtorch/modules/primitive.py (hunk 1 above)
Comment (Guillaumemartinfesta) [reply pyqtorch/modules/primitive.py:461]: Yep you're right !

**Comment by Guillaumemartinfesta** [comment]
> Very nice @Guillaumemartinfesta, I think it looks good :)
> 
> Will you be needing just the multi-qubit toffoli or also other multi-qubit controlled operations?

For now I will only need the multi controlled toffoli but starting from there it should be quite easy to add new multi controlled gates if necessary :)

**Comment by Guillaumemartinfesta** [comment]
> Just took a super brief look but this looks alright to me

So are you okay if I merge ?

**Review by dominikandreasseitz — APPROVED** [review]
great job @Guillaumemartinfesta , i see the others also already reviewed the MR so i only had a brief look but looks exactly like we discussed. we can merge it imo

File: tests/test_modules.py [hunk 4]
```diff
@@ -18,6 +19,19 @@
...
+
+state_0000 = pyq.zero_state(4, device=DEVICE, dtype=DTYPE)
+state_1110 = pyq.X(qubits=[0], n_qubits=4)(
+    pyq.X(qubits=[1], n_qubits=4)(pyq.X(qubits=[2], n_qubits=4)(state_0000))
+)
+state_1111 = pyq.X(qubits=[0], n_qubits=4)(
+    pyq.X(qubits=[1], n_qubits=4)(
+        pyq.X(qubits=[2], n_qubits=4)(pyq.X(qubits=[3], n_qubits=4)(state_0000))
+    )
+)
```
Comment (awennersteen) [inline tests/test_modules.py:35]: Not for this MR, but it would be nice to put these as fixtures in  a conftest!

**Review by awennersteen — APPROVED** [review]

**Pull request merged**